assembly/fs/sync.ts

```diff
@@ -0,0 +1,123 @@
+import {
+  oflags,
+  rights,
+  lookupflags,
+  iovec,
+} from "@assemblyscript/wasi-shim/assembly/bindings/wasi_snapshot_preview1";
+import { UnmanagedResult } from "../error";
+import { OBJECT, TOTAL_OVERHEAD } from "rt/common";
+import { readDirUnsafe, readFileUnsafe, writeFileUnsafe } from "./unsafe";
+import { Dirent, parseOFlags, parseRights } from "./util";
+
+/** Write a file to the filesystem. Encoding can be either "utf8" or "utf16le" */
+export function writeFile<T>(path: string, contents: T, flags: string = "w", encoding: string = "utf8"): UnmanagedResult<usize> {
+  let pathPtr = String.UTF8.encode(path);
+  let pathLen = <usize>pathPtr.byteLength;
+
+  let rights: rights = parseRights(flags);
+  let oflags: oflags = parseOFlags(flags);
+  if (rights == u64.MAX_VALUE || oflags == u16.MAX_VALUE) return new UnmanagedResult<usize>(0, "Invalid flags.");
+
+  if (contents instanceof String) {
+    if (encoding == "utf8" || encoding == "utf-8") {
+      let bytes = String.UTF8.encode(<string>contents);
+      return writeFileUnsafe(
+        changetype<usize>(pathPtr),
+        pathLen,
+        changetype<usize>(bytes),
+        <usize>bytes.byteLength,
+        lookupflags.SYMLINK_FOLLOW,
+        rights,
+        oflags,
+      );
+    } else if (encoding == "utf16le") {
+      return writeFileUnsafe(
+        changetype<usize>(pathPtr),
+        pathLen,
+        changetype<usize>(contents),
+        <usize>changetype<OBJECT>(changetype<usize>(contents) - TOTAL_OVERHEAD).rtSize,
+        lookupflags.SYMLINK_FOLLOW,
+        rights,
+        oflags,
+      );
+    } else return new UnmanagedResult<usize>(0, "Invalid encoding: " + encoding);
+  } else if (contents instanceof ArrayBuffer) {
+    return writeFileUnsafe(
+      changetype<usize>(pathPtr),
+      pathLen,
+      changetype<usize>(contents),
+      <usize>contents.byteLength,
+      lookupflags.SYMLINK_FOLLOW,
+      rights,
+      oflags,
+    );
+    // @ts-ignore: ArrayBufferView is defined
+  } else if (contents instanceof ArrayBufferView) {
+    // @ts-ignore: ArrayBufferView is defined
+    let dataStart: usize = contents.dataStart;
```

Comment:
No need for the variable since it is just renaming `contents.dataStart` to `dataStart`. Just replace `dataStart` with its appropriate value in the `writeFileUnsafe` call @ L60